lib/Push.php

```diff
@@ -112,12 +128,44 @@ public function isDeferring(): bool {
 	}
 
 	public function deferPayloads(): void {
+		$this->deferPreparing = true;
 		$this->deferPayloads = true;
 	}
 
 	public function flushPayloads(): void {
-		$this->deferPayloads = false;
+		$this->deferPreparing = false;
+
+		if (!empty($this->loadDevicesForUsers)) {
+			$missingDevicesFor = array_diff($this->loadDevicesForUsers, array_keys($this->userDevices));
+			$this->loadDevicesForUsers($missingDevicesFor);
+			$this->loadDevicesForUsers = [];
+		}
+
+		if (!empty($this->loadStatusForUsers)) {
+			$missingStatusFor = array_diff($this->loadStatusForUsers, array_keys($this->userStatuses));
+			$newUserStatuses = $this->userStatusManager->getUserStatuses($missingStatusFor);
+			foreach ($missingStatusFor as $userId) {
+				$this->userStatuses[$userId] = $newUserStatuses[$userId] ?? null;
+			}
+			$this->loadStatusForUsers = [];
+		}
+
+		if (!empty($this->notificationsToPush)) {
+			foreach ($this->notificationsToPush as $id => $notification) {
+				$this->pushToDevice($id, $notification);
+			}
+			$this->notificationsToPush = [];
+		}
+
+		if (!empty($this->deletesToPush)) {
+			foreach ($this->deletesToPush as $id => $data) {
+				$this->pushDeleteToDevice($data['userId'], $id, $data['app']);
+			}
+			$this->deletesToPush = [];
+		}
+
 		$this->sendNotificationsToProxies();
+		$this->deferPayloads = false;
```

Comment:
Any special reason why `deferPayloads` is reset now after `sendNotificationsToProxies()`?

Reply:
Yeah because otherwise preparing would not be done again. There are now 2 defers. 1 defers everything (so we first know which users devices and statuses to load) and the other one only the sending part (so we open only one connection to the push proxy)

Reply:
But where is `deferPayloads` used inside `sendNotificationsToProxies()` that now it needs to be reset after it was executed?

Reply:
deferPayloads existed before, usage is at the end of `pushToDevice` and `pushDeleteToDevice`:
```php
		if (!$this->deferPayloads) {
			$this->sendNotificationsToProxies();
		}
```

So yes, I can move it back before `sendNotificationsToProxies` to avoid confusion. It just needs to be after the `pushToDevice` and `pushDeleteToDevice

Reply:
Yes, that is what I meant :-)